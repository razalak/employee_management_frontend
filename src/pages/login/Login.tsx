import { Button, CheckBox } from "../../components";
import { TextInputField } from "../../components";
import kvLogo from "/assets/kv-logo.png";
import kvLoginAside from "/assets/kv-login.jpeg";
import "./Login.css";
import { useEffect, useRef, useState } from "react";
// import useMousePointer from "../../hooks/useMousePointer";
import useLocalStorage from "../../hooks/useLocalStorage";
import { Navigate, useNavigate } from "react-router-dom";


const Login = () => {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  const [validationErrors, setValidationErrors] = useState({
    usernameError: "",
    passwordError: "",
  });
  const usernameRef = useRef<HTMLInputElement | null>(null);
  // const mousePosition = useMousePointer();
  const localStorageHook = useLocalStorage();
  const [isPasswordShown, setIsPasswordShown] = useState(
    Boolean(localStorageHook.get("showPassword"))
  );
   const navigate=useNavigate();
  
 
 const handleLogin=()=>{
    if(username==="dias"&&password==="12345"){
    localStorage.setItem("IsLoggedIn","true");
    navigate("/employee");
  }
}


  useEffect(() => {
    if (username.length > 10) {
      setValidationErrors((err) => {
        return { ...err, usernameError: "Username too long" };
      });
    } else if (!username) {
      setValidationErrors((err) => {
        return { ...err, usernameError: "Username is required" };
      });
    } else {
      setValidationErrors((err) => {
        return { ...err, usernameError: "" };
      });
    }
  }, [username]);

  useEffect(() => {
    if (usernameRef.current) {
      usernameRef.current.focus();
    }
  }, []);


  if(localStorage.getItem("IsLoggedIn")==="true"){
    return <Navigate to="/employee"/>
  }
  return (
    <div className="login-content">
      <aside className="login-aside">
        <img
          className="circle-img"
          src={kvLoginAside}
          alt="stock image for login"
        />
      </aside>

      <main className="login-main">
        <div className="login-form-content">
          <div className="logo-div">
            <a href="#">
              <img className="logo" src={kvLogo} alt="logo" />
            </a>
            {/* <p>{`x: ${mousePosition.x}, y: ${mousePosition.y}`}</p> */}
          </div>
          <div>
            <TextInputField
              placeholder="Username"
              label="Username"
              variants="animated"
              value={username}
              ref={usernameRef}
              onChange={(e) => setUsername(e.target.value)}
              endAdornment={
                <button
                  type="button"
                  onClick={() => setUsername("")}
                  style={{
                    background: "none",
                    border: "none",
                    fontSize: "16px",
                    cursor: "pointer",
                    display: username.length === 0 ? "none" : "block",
                  }}
                >
                  &times;
                </button>
              }
            />
            <p
              style={{
                marginLeft: "6px",
                color: validationErrors.usernameError ? "red" : "green",
              }}
            >
              {validationErrors.usernameError || "Valid username"}
            </p>
          </div>
          <div>
            <TextInputField
              placeholder="Password"
              label="Password"
              type={isPasswordShown ? "text" : "password"}
              variants="animated"
              value={password}
              onChange={(e) => setPassword(e.target.value)}
            />
            <CheckBox
              label="Show password"
              name="show-password"
              checkedState={isPasswordShown}
              onCheckedChange={() => {
                setIsPasswordShown((oldVal) => {
                  localStorageHook.set("showPassword", String(!oldVal));
                  return !oldVal;
                });
              }}
            />
          </div>
          <p>
            {username &&
              `Username: ${
                username + ` ${password && `& password: ${password}`}`
              }`}
          </p>
          <Button
            label="Log In"
            onClick={handleLogin}
            variants="default full-width"
          />
        </div>
      </main>
    </div>
  );
};


export default Login;
