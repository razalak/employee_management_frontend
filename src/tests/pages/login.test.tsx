import { describe, it, expect, beforeEach, vi } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import { Provider } from "react-redux";
import { configureStore } from "@reduxjs/toolkit";
import userEvent from "@testing-library/user-event";
import Login from "../../pages/login/Login";
// Mock react-router-dom's useNavigate
const mockNavigate = vi.fn();
vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});
// Mock the store
const mockStore = configureStore({
  reducer: {
    employee: (state = { employees: [] }) => state,
    department: (state = { departments: [] }) => state,
    employeeApi: (state = {}) => state,
  },
});
// Mock the login mutation
const mockLoginMutation = vi.fn();
vi.mock("../../api-service/auth/login.api", () => ({
  useLoginMutation: () => [
    () => ({
      unwrap: () => mockLoginMutation(),
    }),
    { isLoading: false },
  ],
}));
describe("Login Page", () => {
  const renderLogin = () => {
    return render(
      <Provider store={mockStore}>
        <Login />
      </Provider>
    );
  };
  beforeEach(() => {
    vi.clearAllMocks();
  });
  it("should match snapshot", () => {
    const { container } = renderLogin();
    expect(container).toMatchSnapshot();
  });

  // Test 2: Render login form with all required elements
  it("should render login form with all required elements", () => {
    renderLogin();

    // Check if username input is present
    const usernameInput = screen.getByLabelText("Username");
    expect(usernameInput).toBeInTheDocument();

    // Check if password input is present
    const passwordInput = screen.getByLabelText("Password");
    expect(passwordInput).toBeInTheDocument();

    // Check if submit button is present
    const submitButton = screen.getByRole("button", { name: /log in/i });
    expect(submitButton).toBeInTheDocument();
    expect(submitButton).toHaveAttribute("type", "submit");
  });

  it("should handle successful login", async () => {
    mockLoginMutation.mockResolvedValueOnce({ token: "fake-token" });

    renderLogin();

    // Fill in the form
    const usernameInput = screen.getByLabelText("Username");
    const passwordInput = screen.getByLabelText("Password");
    const submitButton = screen.getByRole("button", { name: /log in/i });

    await userEvent.type(usernameInput, "test@example.com");
    await userEvent.type(passwordInput, "password123");

    // Submit the form
    await userEvent.click(submitButton);

    // Wait for the login mutation to resolve and navigation to occur
    await waitFor(() => {
      expect(mockLoginMutation).toHaveBeenCalled();
      expect(mockNavigate).toHaveBeenCalledWith("/employee");
    });
  });
});

