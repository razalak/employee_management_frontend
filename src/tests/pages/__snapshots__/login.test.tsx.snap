// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Page > should match snapshot 1`] = `
<div>
  <div
    class="login-content"
  >
    <aside
      class="login-aside"
    >
      <img
        alt="stock image for login"
        class="circle-img"
        src="/assets/kv-login.jpeg"
      />
    </aside>
    <main
      class="login-main"
    >
      <div
        class="login-form-content"
      >
        <div
          class="logo-div"
        >
          <a
            href="#"
          >
            <img
              alt="logo"
              class="logo"
              src="/assets/kv-logo.png"
            />
          </a>
        </div>
        <div>
          <div
            class="text-input-field-div"
          >
            <label
              class="label--animated"
            >
              <input
                class="input--animated"
                placeholder=""
                required=""
                type="text"
                value=""
              />
              <span>
                Username
              </span>
              <div
                class="end-adornment"
              >
                <button
                  style="background: none; font-size: 16px; cursor: pointer; display: none;"
                  type="button"
                >
                  Ã—
                </button>
              </div>
            </label>
          </div>
          <p
            style="margin-left: 6px; color: red;"
          >
            Username is required
          </p>
        </div>
        <div>
          <div
            class="text-input-field-div"
          >
            <label
              class="label--animated"
            >
              <input
                class="input--animated"
                placeholder=""
                required=""
                type="password"
                value=""
              />
              <span>
                Password
              </span>
              <div
                class="end-adornment"
              />
            </label>
          </div>
          <label>
            <input
              name="show-password"
              type="checkbox"
            />
            <span>
              Show password
            </span>
          </label>
        </div>
        <p />
        <div
          style="color: red;"
        />
        <button
          class="button  button--default button--full-width undefined"
          type="submit"
        >
          Log In
        </button>
      </div>
    </main>
  </div>
</div>
`;
